include/swift/SILOptimizer/Utils/ConstExpr.h

```diff
@@ -40,6 +42,89 @@ class SymbolicValueAllocator;
 class ConstExprFunctionState;
 class UnknownReason;
 
+class StringLiteralInfo {
+  StringLiteralInfo() = default;
+
+public:
+  bool isAscii;
+  StringRef value;
+
+  static Optional<StringLiteralInfo> create(SILValue v) {
+    if (auto *inst = dyn_cast<ApplyInst>(v))
+      return create(inst);
+    return {};
+  }
+
+  static Optional<StringLiteralInfo> create(SILInstruction *inst) {
+    ApplyInst *makeStr = getStringMakeUTF8Apply(inst);
+    if (!makeStr)
+      return {};
+
+    auto strVal = getUTF8String(makeStr);
+    if (!makeStr)
+      return {};
+
+    StringLiteralInfo info;
+    info.value = strVal.getValue();
+    info.isAscii = getIsAscii(makeStr);
+    return info;
+  }
+
+  /// If the given instruction is a call to the compiler-intrinsic initializer
+  /// of String that accepts string literals, return the called function.
+  /// Otherwise, return nullptr.
+  static SILFunction *getStringMakeUTF8Init(SILInstruction *inst) {
+    auto *apply = dyn_cast<ApplyInst>(inst);
+    if (!apply)
+      return nullptr;
+
+    SILFunction *callee = apply->getCalleeFunction();
+    if (!callee || !callee->hasSemanticsAttr("string.makeUTF8"))
+      return nullptr;
+    return callee;
+  }
+
+  /// Similar to getStringMakeUTF8Init but, gets the apply instruction instead.
+  static ApplyInst *getStringMakeUTF8Apply(SILInstruction *inst) {
+    auto *apply = dyn_cast<ApplyInst>(inst);
+    if (!apply)
+      return nullptr;
+
+    SILFunction *callee = apply->getCalleeFunction();
+    if (!callee || !callee->hasSemanticsAttr("string.makeUTF8"))
+      return nullptr;
+    return apply;
+  }
+
+  /// Gets an optional StringRef from a string init function.
+  static Optional<StringRef> getUTF8String(ApplyInst *makeStr) {
+    SILFunction *callee = makeStr->getCalleeFunction();
+    if (!callee || !callee->hasSemanticsAttr("string.makeUTF8"))
+      return {};
+    
+    if (makeStr->getNumArguments() < 1)
+      return {};
+
+    if (auto *sli = dyn_cast<StringLiteralInst>(makeStr->getOperand(1)))
+      return sli->getValue();
+    return {};
+  }
+
+  /// Gets is string ascii from string init function.
+  static bool getIsAscii(ApplyInst *makeStr) {
```

Comment:
I dislike how this code is assuming that string.makeUTF8 has a certain form. I wonder if we could put in an assert of some sort that checks that any function with string.makeUTF8 that this is called upon has the proper form? Not sure if it is worth it. Definitely though the expected functional form should be documented here.

Otherwise, can you improve this comment along the lines of the one above the StringLiteralInst.

Reply:
I suppose I could assert that the label is correct (`isASCII`) but, I think it's probably OK with a comment on it that describes how the function should look.